dask/array/einsum.py

```diff
@@ -0,0 +1,253 @@
+from __future__ import division, print_function, absolute_import
+
+from distutils.version import LooseVersion
+from functools import wraps
+
+import numpy as np
+from numpy.compat import basestring
+
+from .core import (atop, asarray)
+
+einsum_symbols = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
+einsum_symbols_set = set(einsum_symbols)
+
+
+# This function duplicates numpy's _parse_einsum_input() function
+# See https://github.com/numpy/numpy/blob/master/LICENSE.txt
+# or NUMPY_LICENSE.txt within this directory
+def parse_einsum_input(operands):
```

Comment:
Just to help me understand, was this added here for compat with old versions of NumPy or was this added to workaround conversion (in the original code) of objects to NumPy arrays instead of Dask arrays?

Reply:
@jakirkham `np.einsum` subscript specifications do not necessarily require the output subscripts to be specified and can have ellipses to indicate broadcasting. `parse_einsum_input` replaces and introduces subscripts for ellipses and outputs respectively, allowing the dask einsum wrapper to properly broadcast chunks. Without this, the logic for `np.einsum` functionality would need to be written in the dask wrapper.

Reply:
Sorry I haven't been following this PR very closely. I thought this code was lifted from an existing (perhaps internal) NumPy function. Is that not true?

Reply:
That's exactly right, this was lifted from an internal NumPy function. It handles all of einsum's messy input parsing (broadcasting, optional outputs, multiple argument signatures), which lets us write a really simple implementation for dask.

Reply:
Was it lifted because of the `asarray` call or is there more going on in this copy-paste?

Reply:
It's a private function `_parse_einsum_input()` in NumPy -- not something that we could rely on reliably without forking.

Reply:
Do you think we could encourage NumPy to make it public/semi-public somehow or are there some obstacles to that?

Should add this is not a blocker. Just thinking of reducing our maintenance burden over the long run.

Reply:
> Do you think we could encourage NumPy to make it public/semi-public somehow or are there some obstacles to that?

I reckon thats a good idea. @dgasmith might have some thoughts on this?